### ISSUE TITLE
Rules don't read docs on factory functions

### ISSUE THREAD

**Issue by jscharett** [gajus/eslint-plugin-jsdoc#21]
When defining a 'class' using a factory, the JSDocs are ignored.  For Example:

```
/**
 * @class A
 * @param {Object} config
 */
return widget({
    doWork: function () {}
.
.
.
});
```

This type of paradigm for creating class level functions is used in many libraries, but it looks like the parser ignores these.  Looks like IterateJSDoc only looks at Functions and nothing else.

**Comment by jscharett** [comment]
This also affects rules like check-tag-names, in which if your documenting a non function property, the jsdoc is skipped due to iterateJSDoc's function scope.

**Comment by brettz9** [comment]
I think this can be merged into #214.

**Comment by brettz9** [comment]
Closing in favor of #214

**Comment by brettz9** [comment]
Btw, while I don't expect to adding this support myself, at least for some time, with the new API within `iterateJsdocs.js` one can now more easily specify the other AST types in which a given rule applies, so it shouldn't be too hard to implement (I also recommend the AST explorer for finding out what types exist for a given structure: https://astexplorer.net/ ):

```js
export default iterateJsdoc(({
  report,
  utils
}) => {
...
}, {returns: [
  'ArrowFunctionExpression',
  'ClassDeclaration',
  'FunctionDeclaration',
  'FunctionExpression'
]});
```
